Added SWIFTLINT_REPORT and SWIFTLINT_REPORT_PATH output variables

lawicko opened the pull request:
I have added 2 output variables:

1. SWIFTLINT_REPORT - when this step is finished, this variable contains the swiftlint output. It's useful if you want to pass this output somewhere, for example to a Slack channel. This change should resolve https://github.com/kimdv/bitrise-step-swiftlint/issues/6
2. SWIFTLINT_REPORT_PATH - when this step is finished, this variable contains the path to the file that contains the report. This is useful if you organize your reports somewhere else, for example you upload them to S3 like I do. Thanks to this you can process that file further, however you want. The default file location is `$BITRISE_DEPLOY_DIR/swiftlint_report.$reporter`, so if you use `html` reporter you will end up with a `swiftlint_report.html` in the default Bitrise deploy directory.

To make this work I had to do one important change:
- I removed the `set -ex` from the source file, it's impossible to set multiple environment variables if the script exits as soon as the `swiftlint` exits. I'm not sure if this makes any difference, other than just continuing execution after `swiftlint` returns

Also, I saw that there was another merge request accepted in the meantime, which allowed for linting lines one by one. The functionality that I'm suggesting in this pull request only works in the `all` option. I think the `changed` option may be flawed, I created a bug report https://github.com/kimdv/bitrise-step-swiftlint/issues/8 and https://github.com/kimdv/bitrise-step-swiftlint/issues/9.

Finally, I tried to make sense of the test folder and test config, but it didn't make any sense to me in the current form, so I refactored it a bit so that you can now actually run `bitrise run test` and it's going to run the linting of the file in the `Test` directory using the config in the `Test/TestConfig`.

I have tested these output variables locally and on my project in bitrise.io, and it seems to work fine, I was able to make use of the `SWIFTLINT_REPORT_PATH` to upload the report to my S3 bucket. Please give it a shot and let me know what you think.

Commit:
Added SWIFTLINT_REPORT and SWIFTLINT_REPORT_PATH

kimdv commented on step.yml:
```diff
@@ -93,3 +93,17 @@ inputs:
+    opts:
+      title: A Swiftlint report path
+      summary: The path to the generated Swiftlint report ($BITRISE_DEPLOY_DIR by default). This is useful if you want to have this report in a separate file and do something with it, for example upload to S3.
+      description: The path to the generated Swiftlint report ($BITRISE_DEPLOY_DIR by default). This is useful if you want to have this report in a separate file and do something with it, for example upload to S3.
```
It says that the `$BITRISE_DEPLOY_DIR` is default but you set `null`?

lawicko replied:
Yes, I don't think it's possible to say here what will be the path, because the output file name depends on the reporter that the users selects. However, I will update the summary field to better describe this.

kimdv commented on step.sh:
```diff
@@ -34,7 +33,14 @@ case $lint_range in
-  swiftlint lint --reporter "${reporter}" ${FLAGS}
+"all")
+  output="$(swiftlint lint --reporter "${reporter}" ${FLAGS})"
+  envman add --key "SWIFTLINT_REPORT" --value "${output}"
+  echo "Saved swiftlint output in SWIFTLINT_REPORT"
+  filename="swiftlint_report.${reporter}"
+  report_path="${BITRISE_DEPLOY_DIR}/${filename}"
+  echo "${output}" > $report_path
+  envman add --key "SWIFTLINT_REPORT_PATH" --value "${report_path}"
+  echo "Saved swiftlint output in file, it's path is saved in SWIFTLINT_REPORT_PATH"
```
Does it only work if it lints all file? Nothing for the diff?

lawicko replied:
Yes, I suspect that the diff functionality doesn't work as expected. Please have a look a the description of this PR, I put there some details on what I think is broken, I have also created 2 bug reports.

kimdv commented on step.sh:
```diff
@@ -34,7 +33,14 @@ case $lint_range in
+  output="$(swiftlint lint --reporter "${reporter}" ${FLAGS})"
+  envman add --key "SWIFTLINT_REPORT" --value "${output}"
+  echo "Saved swiftlint output in SWIFTLINT_REPORT"
+  filename="swiftlint_report.${reporter}"
```
This will not work correctly I think. We cannot have a file called `.xcode` etc 😅

Maybe this could help?
https://github.com/limamedeiros/bitrise-step-swiftlint/blob/master/step.sh#L19

lawicko replied:
This helps a lot, thanks! I will update this part soon.

kimdv reviewed (changes requested):
Thanks a lot! 🚀 🙏

I have a few comments and suggestions. Let me hear what you think!

Commit:
Improved the report file names by using proper extensions for all reporter types, improved the SWIFTLINT_REPORT_PATH to describe how it is created

Commit:
A typo in the field summary

kimdv commented on step.yml:
```diff
@@ -93,3 +93,17 @@ inputs:
       - "yes"
       - "no"
 outputs:
+  - SWIFTLINT_REPORT: null
+    opts:
+      title: A Swiftlint report
+      summary: The generated Swiftlint report. May be empty if the Swiftlint succeeded with 0 errors or warnings. This is useful when you want to pass it somewhere else, for example on a Slack channel.
+      description: The generated Swiftlint report. May be empty if the Swiftlint succeeded with 0 errors or warnings. This is useful when you want to pass it somewhere else, for example on a Slack channel.
+      is_dont_change_value: true
+      category: Output
+  - SWIFTLINT_REPORT_PATH: null
+    opts:
+      title: A Swiftlint report path
+      summary: The path to the generated Swiftlint report (by default, it is created by concatenating the $BITRISE_DEPLOY_DIR with the generated report file name. The file name depends on the reporter that was selected, for example for the `junit` reporter, the generated file name will be `swiftlint_report.xml`, and for the `html` reporter it will be `swiftlint_report.html`). This is useful if yo ...
+      description: The path to the generated Swiftlint report (by default, it is created by concatenating the `$BITRISE_DEPLOY_DIR` with the generated report file name. The file name depends on the reporter that was selected, for example for the `junit` reporter, the generated file name will be `swiftlint_report.xml`, and for the `html` reporter it will be `swiftlint_report.html`). This is useful ...
+      is_dont_change_value: true
+      category: Output
```
It will only create one output file. The `SWIFTLINT_REPORT_PATH ` should be an input I think. 
And if it's always `BITRISE_DEPLOY_DIR` then we should not comment on this.

kimdv committed:
Made some simplifications 

Redirected the `stdout` from SwiftLint based on https://github.com/realm/SwiftLint/issues/2977

kimdv committed:
Rearranged echo and SwiftLint

kimdv commented:
@lawicko I have pushed some small changes, hope it's okay 🤓

lawicko commented:
> @lawicko I have pushed some small changes, hope it's okay 🤓

The lines starting with `envman` were the core of the new functionality, when you deleted them, now those variables are not created and it will not work anymore 😅

kimdv commented:
Hmm I get a file when I run locally in the deploy path.
Will read the other one again.

lawicko commented:
> Hmm I get a file when I run locally in the deploy path.

Sure, you will get the file locally, but you have no way to reference that file on Bitrise,  because you removed the lines that were actually creating these environment variables. You also killed the variable where the output of the swiftlint was kept, so that functionality is gone altogether.

I suggest you revert your changes and tell me what are you trying to achieve, and we can take it from there.

kimdv committed:
Readded SWIFTLINT_REPORT 🤦‍♂️

kimdv committed:
Correct write

kimdv committed:
Update step.sh

lawicko commented:
@kimdv Look, I'm already lost in the changes you made, let's do this again, this time just me writing code and you reviewing.

kimdv commented:
I cannot see why we use `SWIFTLINT_REPORT_PATH` envman to write to a file? 
Seems not correct to me.. :)

lawicko commented:
> I cannot see why we use `SWIFTLINT_REPORT_PATH` envman to write to a file?
> Seems not correct to me.. :)

This is not writing to a file, this is creating the environment variable which can later be used on Bitrise to further process that file. The file is created here: `echo "${output}" > $report_path`
